### PR TITLE
Separate recId calc from signing

### PR DESCRIPTION
**Pull request by rain-on**
## PR description
Security frameworks create signatures of 64-bytes (for a 256-bit keys) - and do not produce the recId.
Furthermore, the signatures created maybe in the upper-half of the curve, which is illegal under Ethereum.

As these checks must be applied to signatures generated by the jce, it is prudent to separate the concept of signature generation, from the post-operations.

## Fixed Issue(s)
N/A

### CODE REVIEW

File: crypto/src/main/java/org/hyperledger/besu/crypto/SECP256K1.java [hunk 1]
```diff
@@ -187,9 +187,17 @@ public static Signature sign(final Bytes32 dataHash, final KeyPair keyPair) {
...
-    BigInteger s = components[1];
 
+    return normaliseSignature(components[0], components[1], keyPair.getPublicKey(), dataHash);
+  }
+
+  public static Signature normaliseSignature(
+      final BigInteger nativeR,
+      final BigInteger nativeS,
+      final PublicKey publicKey,
+      final Bytes32 dataHash) {
```
Comment (jframe) [inline crypto/src/main/java/org/hyperledger/besu/crypto/SECP256K1.java:198]: DataHash shouldn't be needed here

Comment (rain-on) [reply crypto/src/main/java/org/hyperledger/besu/crypto/SECP256K1.java:198]: its used to ensure the foward-backward signature works

**Comment by shemnon** [comment]
So this is kind of concerning since signature recovery depends on having the recovery Id.  How do we deal with that?

**Comment by rain-on** [comment]
So - we still do the recovery - and the "lower half curve" - however, when you use other security frameworks (eg via jce) does _not_ provide the recId - only the r & s.

Thus, if we move to other security providers, the responses from them must go through the "half curve" and "recId" processes - thus that aspect of signing needs to be separate function - the signing function in Besu is unchanged.

**Pull request merged**